src/functional/iterator/pretty_printer.py

```diff
@@ -0,0 +1,277 @@
+                if (
+                    op == "*"
+                    and isinstance(node.args[0], ir.FunCall)
+                    and isinstance(node.args[0].fun, ir.FunCall)
+                    and node.args[0].fun.fun == ir.SymRef(id="shift")
+                    and node.args[0].fun.args
+                    and isinstance(node.args[0].args[0], ir.SymRef)
+                ):
+                    # deref(shift(offsets...)(sym)) → sym[offsets...]
+                    assert len(node.args[0].args) == 1
+                    expr = self.visit(node.args[0].args[0], prec=OTHER_PRECEDENCE["call"])
+                    shifts = self.visit(node.args[0].fun.args, prec=0)
+                    res = _hmerge(expr, ["["], *_hinterleave(shifts, ", "), ["]"])
+                    return _prec_parens(res, prec, OTHER_PRECEDENCE["call"])
+                res = _hmerge([op], self.visit(node.args[0], prec=UNARY_PRECEDENCE[op]))
+                return _prec_parens(res, prec, UNARY_PRECEDENCE[op])
+            if node.fun.id == "tuple_get" and len(node.args) == 2:
+                # tuple_get(i, x) → x[i]
+                idx, tup = self.visit(node.args, prec=OTHER_PRECEDENCE["call"])
+                res = _hmerge(tup, ["["], idx, ["]"])
+                return _prec_parens(res, prec, OTHER_PRECEDENCE["call"])
+            if node.fun.id == "named_range" and len(node.args) == 3:
+                # named_range(dim, start, stop) → dim: [star, stop)
+                dim, start, end = self.visit(node.args, prec=0)
+                res = _hmerge(dim, [": ["], start, [", "], end, [")"])
+                return _prec_parens(res, prec, OTHER_PRECEDENCE["call"])
+            if node.fun.id == "domain" and len(node.args) >= 1:
+                # domain(x, y, ...) → { x × y × ... }
+                args = self.visit(node.args, prec=OTHER_PRECEDENCE["call"])
+                return _hmerge(["{ "], *_hinterleave(args, " × "), [" }"])
+            if node.fun.id == "if_" and len(node.args) == 3:
+                # if_(x, y, z) → if x then y else z
+                ifb, thenb, elseb = self.visit(node.args, prec=OTHER_PRECEDENCE["if"])
+                hblocks = _hmerge(["if "], ifb, [" then "], thenb, [" else "], elseb)
+                vblocks = _vmerge(
+                    _hmerge(["if   "], ifb), _hmerge(["then "], thenb), _hmerge(["else "], elseb)
+                )
+                return _prec_parens(_optimum(hblocks, vblocks), prec, OTHER_PRECEDENCE["if"])
+
+        fun = self.visit(node.fun, prec=OTHER_PRECEDENCE["call"])
+        args = self.visit(node.args, prec=0)
+
+        if not args:
+            args = [""]
+        else:
+            hargs = _hmerge(*_hinterleave(args, ", "))
+            vargs = _vmerge(*_hinterleave(args, ","))
+            args = _optimum(hargs, vargs)
+
+        if node.fun == ir.SymRef(id="make_tuple"):
+            # make_tuple(args...) → {args...}
+            htup = _hmerge(["{"], args, ["}"])
```

Comment:
Just a question: would it make sense to define separators (except parenthesis) and tokens in a common place and reuse them later in the methods. It simplifies customization and changes in the beautified format itself, but it might be overkill for our use cases.

Reply:
Hmm, probably that’s an overkill indeed. It’s also tricky to customize the horizontal/vertical merging etc. Probably it’s just easiest to modify things in place.